sources/OpenMcdf.Extensions/OLEProperties/PropertyFactory.cs

```diff
@@ -412,26 +412,86 @@ public override string ReadScalarValue(System.IO.BinaryReader br)
                 uint size = br.ReadUInt32();
                 data = br.ReadBytes((int)size);
 
-                return Encoding.GetEncoding(codePage).GetString(data);
+                var result = Encoding.GetEncoding(codePage).GetString(data);
+                //result = result.Trim(new char[] { '\0' });
+
+                //if (this.codePage == CodePages.CP_WINUNICODE)
+                //{
+                //    result = result.Substring(0, result.Length - 2);
+                //}
+                //else
+                //{
+                //    result = result.Substring(0, result.Length - 1);
+                //}
+
+                return result;
             }
 
             public override void WriteScalarValue(BinaryWriter bw, string pValue)
             {
-                data = Encoding.GetEncoding(codePage).GetBytes(pValue);
-                uint dataLength = (uint)data.Length;
+                //bool addNullTerminator = true;
+
+                if (String.IsNullOrEmpty(pValue)) //|| String.IsNullOrEmpty(pValue.Trim(new char[] { '\0' })))
+                {
+                    bw.Write((uint)0);
+                }
+                else if (this.codePage == CodePages.CP_WINUNICODE)
+                {
 
-                // The string data must be null terminated, so add a null byte if there isn't one already
-                bool addNullTerminator =
-                    dataLength == 0 || data[dataLength - 1] != '\0';
+                    data = Encoding.GetEncoding(codePage).GetBytes(pValue);
+
+                    //if (data.Length >= 2 && data[data.Length - 2] == '\0' && data[data.Length - 1] == '\0')
+                    //    addNullTerminator = false;
+
+                    uint dataLength = (uint)data.Length;
+
+                    //if (addNullTerminator)
+                    dataLength += 2;            // null terminator \u+0000
+
+                   // var mod = dataLength % 4;       // pad to multiple of 4 bytes
+
+                    bw.Write(dataLength);           // datalength of string + null char (unicode)
+                    bw.Write(data);                 // string
+
+
+                    //if (addNullTerminator)
+                    //{
+                    bw.Write('\0');                 // first byte of null unicode char
+                    bw.Write('\0');                 // second byte of null unicode char
+                    //}

```

Comment:
So for writing the null terminator + padding, what about this for non-unicode LPSTR ... 
```
byte[] zeroes = [0, 0, 0, 0];
data = Encoding.GetEncoding(codePage).GetBytes(pValue);
uint dataLength = (uint)data.Length + 1; // Add null terminator to length
bw.Write(dataLength);
bw.Write(data);

dataLength = ((4 - (dataLength % 4)) % 4) + 1; // determine padding plus null terminator
bw.Write(zeroes, 0, (int)dataLength);
```
This way you're not issuing individual writes for each null padding, and it includes the null terminator.

For unicode LPWSTR just needs minimal modification:
```
byte[] zeroes = [0, 0, 0, 0];
data = Encoding.Unicode.GetBytes(pValue);
uint dataLength = (uint)data.Length + 2; // Add 2-byte null terminator
bw.Write(dataLength / 2);
bw.Write(data);
dataLength = ((4 - (dataLength % 4)) % 4) + 2; // Determine padding plus 2-byte null terminator
bw.Write(zeroes, 0, (int)dataLength);
```

Reply:
Padding is generalized in Parent method `TypedPropertyValue.Write` for manteinabilty.